### ISSUE TITLE
Test listener overwrites `test_ids` file when executed in parallel. 

### ISSUE THREAD

**Issue by lazar-mitrovic** [graalvm/native-build-tools#84]
When invoking `test` task with large amounts of tests, multiple JVM processes are being spawned in parallel. When test listener is triggered, all of those processes overwrite the same file (instead of appending to it). This results in `test_ids` file which contains only the tests that were executed in the last JVM process (which is undeterministic and might result in a empty file). 

This behavior can be triggered by invoking the test task with 500+ tests on Gradle plugin.

Reproducer:
```bash
git clone git@github.com:rsocket/rsocket-java.git
curl https://gist.githubusercontent.com/lazar-mitrovic/b075423362aef58c2d7d1b4132e6899e/raw/cb4ef77a4989b1985c2d1f999136037e40acba63/rsocket-repro.diff -O
cd rsocket-java
git apply ../rsocket-repro.diff
./gradlew clean rsocket-core:test rsocket-core:nativeTest
```
Issue dissapears when working with a smaller number of tests - uncomment the test filter from the end of `rsocket-core/build.gradle` and it should work as intended.

AFAIK this issue is also present in listener from the current [JUnit5 master](https://github.com/junit-team/junit5/blob/main/junit-platform-launcher/src/main/java/org/junit/platform/launcher/listeners/UniqueIdTrackingListener.java).

Possible fix might be to separate listener output files, or to introduce some sort of locking mechanism that would ensure that file is accessed correctly.

Potential workaround for the end-users ATM might be to use some Gradle magic to force all tasks to be executed sequentially and in the same JVM, but this would come with a extreme performance penalty (and I'm unsure if such configuration option is even present).

cc: @sbrannen @gradinac @melix

**Comment by melix** [comment]
That's an interesting discovery indeed. This could potentially be fixed by implementing this "dry run" mode (#76), granted that discovery is done in a single process.

Alternatively, the test listener could, instead of a file, be passed an output directory. Then:

- the build tool would be responsible for passing a "clean" directory when the tests are started
- the listener would be responsible for generating a unique file in that directory (say `test_ids.txt` exist, create `test_ids2.txt`, ...)
- the build tool would merge the test ids files into a single one

In #71 I already added the ability to pass in the path to the generated file, but we could use a _directory_ instead. That would probably be cheaper/more reliable than the dry run for now.

**Comment by sbrannen** [comment]
Related changes to the official `UniqueIdTrackingListener` coming in JUnit 5.8 have been made in https://github.com/junit-team/junit5/commit/9a1db7309f908f474b6042c9eb68c1e596e58453.
